src/NuGet.Core/NuGet.CommandLine.XPlat/CollectorLogger.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using NuGet.Logging;
+
+namespace NuGet.CommandLine.XPlat
+{
+    public class CollectorLogger : ICollectorLogger
+    {
+        private readonly ILogger _innerLogger;
+
+        /// <summary>
+        /// Each log level that wrill be recorded has an object (for locking) and a list of log
+        /// messages.
+        /// </summary>
+        private readonly IDictionary<LogLevel, LockAndList> _logLists;
+
+        public CollectorLogger(ILogger innerLogger)
+        {
+            _innerLogger = innerLogger;
+
+            _logLists = new Dictionary<LogLevel, LockAndList>
+            {
+                { LogLevel.Error, new LockAndList() }
+            };
+        }
+        
+        public void LogDebug(string data)
+        {
+            _innerLogger.LogDebug(data);
+        }
+
+        public void LogVerbose(string data)
+        {
+            _innerLogger.LogVerbose(data);
+        }
+
+        public void LogInformation(string data)
+        {
+            _innerLogger.LogInformation(data);
+        }
+
+        public void LogWarning(string data)
+        {
+            _innerLogger.LogWarning(data);
+        }
+
+        public void LogError(string data)
+        {
+            AddLog(LogLevel.Error, data);
+
+            _innerLogger.LogError(data);
+        }
+
+        /// <summary>
+        /// Fetch all of the errors logged since that last time this method was called. This method
+        /// is useful when error log messages should be redisplayed after the initial log message
+        /// is emitted.
+        /// </summary>
+        public IEnumerable<string> FlushErrors()
+        {
+            return FlushLogs(LogLevel.Error);
+        }
+
+        private void AddLog(LogLevel logLevel, string data)
+        {
+            LockAndList lockAndList;
+            if (!_logLists.TryGetValue(logLevel, out lockAndList))
+            {
+                return;
+            }
+
+            lock (lockAndList.Lock)
```

Comment:
This may really slow things down. Could we just remove the lock and make sure that in xplat the tasks have finished before reading and clearing?